Google.Solutions.Compute/Extensions/ResetWindowsUser.cs

```diff
@@ -0,0 +1,147 @@
+{
+    public static class ResetWindowsUserExtensions
+    {
+        private const int RsaKeySize = 2048;
+        private const int SerialPort = 4;
+        private const string MetadataKey = "windows-keys";
+
+        public static Task<NetworkCredential> ResetWindowsUserAsync(
+            this InstancesResource resource,
+            string project,
+            string zone,
+            string instance,
+            string username)
+        {
+            return ResetWindowsUserAsync(
+                resource,
+                new VmInstanceReference(project, zone, instance),
+                username);
+        }
+
+        public static async Task<NetworkCredential> ResetWindowsUserAsync(
+            this InstancesResource resource,
+            VmInstanceReference instanceRef,
+            string username)
+        {
+            using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider(RsaKeySize))
+            {
+                var keyParameters = rsa.ExportParameters(false);
+
+                var requestPayload = new RequestPayload()
+                {
+                    ExpireOn = DateTime.UtcNow.AddMinutes(5),
+                    Username = username,
+                    Email = username,
+                    Modulus = Convert.ToBase64String(keyParameters.Modulus),
+                    Exponent = Convert.ToBase64String(keyParameters.Exponent),
+                };
+
+                // Send the request to the instance via a special metadata entry.
+                await resource.AddMetadataAsync(
+                    instanceRef,
+                    MetadataKey,
+                    JsonConvert.SerializeObject(requestPayload)).ConfigureAwait(false);
+
+                // Read response from serial port.
+                for (int attempt = 0; attempt < 40; attempt++)
+                {
+                    string buffer = await resource.GetSerialPortOutputStream(
+                        instanceRef, 
+                        SerialPort).ReadAsync().ConfigureAwait(false);
+                    var response = buffer.Split('\n')
+                        .Where(line => line.Contains(requestPayload.Modulus))
+                        .FirstOrDefault();
+                    if (response == null)
+                    {
+                        await Task.Delay(500).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        var responsePayload = JsonConvert.DeserializeObject<ResponsePayload>(response);
+                        if (!string.IsNullOrEmpty(responsePayload.ErrorMessage))
+                        {
+                            throw new PasswordResetException(responsePayload.ErrorMessage);
+                        }
+
+                        var password = rsa.Decrypt(
+                            Convert.FromBase64String(responsePayload.EncryptedPassword),
+                            true);
+
+                        return new NetworkCredential(
```

Comment:
just note: w'ere encrypting a secret here and saving it within windows context...which is fine  if its saved securely.

i understand that this allows for you to use this without gcloud at all..


however, gcloud does now in alpa support `gcloud alpha secrets get/set` which will allow you to read/set asecret (eg, any password)....but this also couples you into using gcloud cli..

Reply:
This is a c# implementation of https://cloud.google.com/compute/docs/instances/windows/automate-pw-generation. The library does not save the resulting password anywhere.